alacritty_terminal/src/term/mod.rs

```diff
@@ -1995,32 +1995,81 @@ impl<T: EventListener> Handler for Term<T> {
     fn graphics_attribute(&mut self, pi: u16, pa: u16) {
         // From Xterm documentation:
         //
+        //   CSI ? Pi ; Pa ; Pv S
+        //
         //   Pi = 1  -> item is number of color registers.
         //   Pi = 2  -> item is Sixel graphics geometry (in pixels).
+        //   Pi = 3  -> item is ReGIS graphics geometry (in pixels).
         //
         //   Pa = 1  -> read attribute.
+        //   Pa = 2  -> reset to default.
+        //   Pa = 3  -> set to value in Pv.
         //   Pa = 4  -> read the maximum allowed value.
         //
-        // Any other request reports an error.
+        //   Pv is ignored by xterm except when setting (Pa == 3 ).
+        //   Pv = n <- A single integer is used for color registers.
+        //   Pv = width ; height <- Two integers for graphics geometry.
+        //
+        //   xterm replies with a control sequence of the same form:
+        //
+        //   CSI ? Pi ; Ps ; Pv S
+        //
+        //   where Ps is the status:
+        //   Ps = 0  <- success.
+        //   Ps = 1  <- error in Pi.
+        //   Ps = 2  <- error in Pa.
+        //   Ps = 3  <- failure.
+        //
+        //   On success, Pv represents the value read or set.
 
-        let (ps, pv) = if pa == 1 || pa == 4 {
-            match pi {
-                1 => (0, &[sixel::MAX_COLOR_REGISTERS][..]),
-                2 => (0, &MAX_GRAPHIC_DIMENSIONS[..]),
-                _ => (1, &[][..]), // Report error in Pi
+        fn generate_response(pi: u16, ps: u16, pv: &[usize]) -> String {
+            let mut text = format!("\x1b[?{};{}", pi, ps);
+            for item in pv {
+                let _ = write!(&mut text, ";{}", item);
+            }
+            text.push('S');
+            text
+        }
+
+        let (ps, pv) = match pi {
+            1 => {
+                match pa {
+                    1 => (3, &[][..]), // Report unsupported
```

Comment:
Why is `Pi = 1, Pa = 2` unsupported?

Right now, it returns the same value of `Pa = 4`.

```console
$ read -p $'\e[?1;1;0S' -rsdS R ; printf '%q\n' "$R"
'\E[?1;0;1024'
```

Reply:
According to the following xterm codes, it is seems that `Pi = 1, Pa = 1` should return current number of color registers if the terminal id is well-known or maximum of number of color registers if the terminal id is unknown:

* https://github.com/ThomasDickey/xterm-snapshots/blob/82e56295226502cae94e943c3f40366f742bb266/charproc.c#L4603-L4606
* https://github.com/ThomasDickey/xterm-snapshots/blob/82e56295226502cae94e943c3f40366f742bb266/graphics.c#L587-L631

The reason why `Pi = 1, Pa = 1` unsupported is because alacritty returns well-known "VT102" for "Send Device Attributes" but it is not easy for current sixel implementation to return current number of color registers.

Reply:
> but it is not easy for current sixel implementation to return current number of color registers.

Xterm has different responses because it allows to set a lower number of color registers, but in our case, the current value is always the maximum (i.e., the default). I think that, in our case, there is no point of differentiate “current” and “maximum” number of color registers.

This relation between current/maximum can be replicated in xterm if we reset the number of registers:

```bash
$ echo $XTERM_VERSION
XTerm(366)

$ read -p $'\e[c' -rsdc R ; printf '%q\n' "$R"
$'\E[?63;1;2;4;6;9;15;16;22;28'

# Initially, the value is set to `256`
$ read -p $'\e[?1;1;0S' -rsdS R ; printf '%q\n' "$R"
$'\E[?1;0;256'

# Reset to default
$ read -p $'\e[?1;2;0S' -rsdS R ; printf '%q\n' "$R"
$'\E[?1;0;1024'

# Now, it is the maximum value.
$ read -p $'\e[?1;1;0S' -rsdS R ; printf '%q\n' "$R"
$'\E[?1;0;1024'
```

Programs like lsix [use this information](https://github.com/hackerb9/lsix/blob/e49fc61e8ccd456add6861fa23fcbb88c5c13752/lsix#L123-L126) to decide how many colors are available in the terminal. I'm worried that if we just remove support for `Pi = 1, Pa = 1` some programs may be broken.

Reply:
Thank you for explanation. I got it.